### PR TITLE
[AIRFLOW-4002] Optionally open debugger on errors in airflow test.

### PR DESCRIPTION
**Pull request by jmcarp**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-4002
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

**Review by potiuk — COMMENTED** [review]
👍 👍 👍 👍 👍

**Comment by potiuk** [comment]
Super useful!

**Review by potiuk — APPROVED** [review]

**Comment by mik-laj** [comment]
I recommend coworkers to cherry-pick this change. This is very helpful in the development of operators.

**Review by mik-laj — APPROVED** [review]

**Comment by mik-laj** [comment]
Have you thought about ipdb support?  Ipdb provides a much more enjoyable experience.  It is possible to detect ipdb. When available, use it. Otherwise, use idpb

**Comment by jmcarp** [comment]
Added `ipdb` detection. Ready for review when you have time @ashb @XD-DENG @feng-tao.

File: airflow/bin/cli.py [hunk 1]
```diff
@@ -2032,6 +2047,11 @@ class CLIFactory(object):
         'task_params': Arg(
             ("-tp", "--task_params"),
             help="Sends a JSON params dict to the task"),
+        'post_mortem': Arg(
+            ("-pm", "--post_mortem"),
+            action="store_true",
```
Comment (XD-DENG) [inline airflow/bin/cli.py:2048]: Seems the default value here would be `True`, i.e. the post_mortem will be invoked by default. But this is different from the previous behavior. Personally I would suggest to use `False` as default value here, to make behavior consistent.

File: airflow/bin/cli.py [hunk 2]
```diff
@@ -686,10 +687,24 @@ def test(args, dag=None):
...
+            debugger.post_mortem()
+        else:
+            raise
+
+
+def get_debugger():
+    try:
+        return importlib.import_module("ipdb")
+    except ImportError:
+        return importlib.import_module("pdb")
```
Comment (XD-DENG) [inline airflow/bin/cli.py]: Do we need to have a separate function here?

I understand this is very "personal preference": but I think we can simply put this part after `if args.post_mortem:` in the `try-except` block above, given the logic inside this `get_debugger()` function is too simple.

**Review by XD-DENG — COMMENTED** [review]
My 2 cents

File: airflow/bin/cli.py (hunk 2 above)
Comment (jmcarp) [reply airflow/bin/cli.py]: Agreed, I factored this back into the `test` function.

File: airflow/bin/cli.py (hunk 1 above)
Comment (jmcarp) [reply airflow/bin/cli.py:2048]: In argparse, store_true sets the default to false, so I think this is correct.

Comment (XD-DENG) [reply airflow/bin/cli.py:2048]: Yes you're right. Thanks for clarification.

File: airflow/bin/cli.py [hunk 3]
```diff
@@ -686,10 +687,20 @@ def test(args, dag=None):
...
+            ti.dry_run()
+        else:
+            ti.run(ignore_task_deps=True, ignore_ti_state=True, test_mode=True)
+    except Exception:
+        if args.post_mortem:
+            try:
+                debugger = importlib.import_module("ipdb")
+            except ImportError:
+                debugger = importlib.import_module("pdb")
+            debugger.post_mortem()
```
Comment (ashb) [inline airflow/bin/cli.py:701]: Oh nice, I didn't know about this function. (I've always just used `.set_trace()`

**Comment by jmcarp** [comment]
Anything else to revise here? I think I addressed everyone's feedback 🤞

**Comment by jmcarp** [comment]
PTAL when you have time @ashb @XD-DENG.

**Pull request merged**